### PR TITLE
Add support for shared fonts in packages.

### PR DESCRIPTION
**Pull request by szakarias**

### CODE REVIEW

File: packages/flutter_tools/lib/src/asset.dart [hunk 1]
```diff
@@ -367,18 +385,74 @@ DevFSContent _createAssetManifest(Map<_Asset, List<_Asset>> assetVariants) {
...
+    if (packageName == null) {
+      fonts.addAll(manifestDescriptor['fonts']);
+    } else {
+      fonts.addAll(_parsePackageFonts(
+        manifestDescriptor['fonts'],
+        packageName,
+        packageMap,
+      ));
+    }
+  return fonts;
```
Comment (mravn-google) [inline packages/flutter_tools/lib/src/asset.dart]: This method no longer returns `null` if the fonts map is empty. Does that work with all callers?

File: packages/flutter_tools/lib/src/asset.dart [hunk 2]
```diff
@@ -367,18 +385,74 @@ DevFSContent _createAssetManifest(Map<_Asset, List<_Asset>> assetVariants) {
...
+}
+
+/// Prefixes family names and asset paths of fonts included from packages with
+/// 'packages/<package_name>
+///
+/// (e.g., replace {"fonts":[{"asset":"a/bar"}],"family":"bar"} by
+/// {"fonts":[{"asset":"packages/foo/a/bar"}],"family":"packages/foo/bar"})
+List<Map<String, dynamic>> _parsePackageFonts(
```
Comment (mravn-google) [inline packages/flutter_tools/lib/src/asset.dart]: Cf. earlier comment. This method might benefit from an explicit `FontDescriptor` class. Hopefully we wouldn't then have to throw around so many variables of type `Map<String, dynamic>` in one place.

File: packages/flutter_tools/lib/src/asset.dart [hunk 3]
```diff
@@ -367,18 +385,74 @@ DevFSContent _createAssetManifest(Map<_Asset, List<_Asset>> assetVariants) {
...
+  final List<Map<String, dynamic>> parsedFonts = <Map<String, dynamic>>[];
+  for (Map<String, dynamic> fontFamily in manifestDescriptor) {
+    final Map<String, dynamic> parsedFontFamily = <String, dynamic>{};
+    for (String key in fontFamily.keys) {
+      if (key == 'family') {
+        parsedFontFamily[key] = 'packages/$packageName/${fontFamily[key]}';
+      } else {
+        assert(key == 'fonts');
+        final List<Map<String, dynamic>> parsedAssets = <Map<String, dynamic>>[];
+        for(Map<String, dynamic> assetProperties in fontFamily[key]) {
```
Comment (mravn-google) [inline packages/flutter_tools/lib/src/asset.dart]: Add space after `for`.

File: packages/flutter_tools/lib/src/asset.dart [hunk 4]
```diff
@@ -367,18 +385,74 @@ DevFSContent _createAssetManifest(Map<_Asset, List<_Asset>> assetVariants) {
...
+          final Map<String, dynamic> parsedAssetProperties = <String, dynamic>{};
+          for (String property in assetProperties.keys) {
+            if (property == 'asset') {
+              final String assetPath = assetProperties[property];
+              if (assetPath.startsWith('packages') &&
```
Comment (mravn-google) [inline packages/flutter_tools/lib/src/asset.dart]: [Avoid curly braces around single-line blocks](https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#avoid-using-braces-for-one-line-long-control-structure-statements).

File: packages/flutter_tools/test/asset_bundle_package_fonts_test.dart [hunk 5]
```diff
@@ -0,0 +1,293 @@
...
+          '"family":"packages/test_package/foo"}]';
+      await buildAndVerifyFonts(
+        <String>[],
+        <String>[font],
+        <String>['test_package'],
+        expectedFontManifest,
+      );
+    }, overrides: contextOverrides);
+
+    testUsingContext('App uses package font with a font file from another package.', () async {
```
Comment (mravn-google) [inline packages/flutter_tools/test/asset_bundle_package_fonts_test.dart]: Drop `a`.

File: packages/flutter_tools/test/asset_bundle_package_fonts_test.dart [hunk 6]
```diff
@@ -0,0 +1,293 @@
...
+        <String>[font],
+        <String>[font],
+        <String>['test_package'],
+        expectedFontManifest,
+      );
+    }, overrides: contextOverrides);
+  });
+}
+
+Map<Type, Generator> get contextOverrides =>
```
Comment (mravn-google) [inline packages/flutter_tools/test/asset_bundle_package_fonts_test.dart]: Can't this be on one line? We should not use `=>` with multi-line syntax.

File: packages/flutter_tools/lib/src/asset.dart (hunk 1 above)
Comment (szakarias) [reply packages/flutter_tools/lib/src/asset.dart]: I believe so.

File: packages/flutter_tools/lib/src/asset.dart [hunk 7]
```diff
@@ -127,18 +128,30 @@ class AssetBundle {
           continue;
         final int result = await _validateFlutterManifest(packageManifest);
         if (result == 0) {
-          final Map<String, dynamic> packageManifestDescriptor = packageManifest;
+          Map<String, dynamic> packageManifestDescriptor = packageManifest;
```
[the remark this answers is not in the document]
Comment (szakarias) [reply packages/flutter_tools/lib/src/asset.dart]: Agree. I will do this in another PR.

File: packages/flutter_tools/test/asset_bundle_package_fonts_test.dart (hunk 6 above)
Comment (szakarias) [reply packages/flutter_tools/test/asset_bundle_package_fonts_test.dart]: It's quite long, changed it to use`{}`.

File: packages/flutter/test/painting/text_style_test.dart [hunk 8]
```diff
@@ -129,5 +129,17 @@ void main() {
     final ui.ParagraphStyle ps7 = const TextStyle().getParagraphStyle(textDirection: TextDirection.rtl);
     expect(ps7, equals(new ui.ParagraphStyle(textDirection: TextDirection.rtl)));
     expect(ps7.toString(), 'ParagraphStyle(textAlign: unspecified, textDirection: TextDirection.rtl, fontWeight: unspecified, fontStyle: unspecified, maxLines: unspecified, fontFamily: unspecified, fontSize: unspecified, lineHeight: unspecified, ellipsis: unspecified)');
+
+    final TextStyle s6 =  const TextStyle(fontFamily: 'test',);
+    expect(s6.fontFamily, 'test');
+    expect(s6.package, isNull);
+    expect(s6.getTextStyle().toString(), 'TextStyle(color: unspecified, decoration: unspecified, decorationColor: unspecified, decorationStyle: unspecified, fontWeight: unspecified, fontStyle: unspecified, textBaseline: unspecified, fontFamily: test, fontSize: unspecified, letterSpacing: unspecified, wordSpacing: unspecified, height: unspecified)');
+
+    final TextStyle s7 =  const TextStyle(fontFamily: 'test', package: 'p',);
```
[the remark this answers is not in the document]
Comment (szakarias) [reply packages/flutter/test/painting/text_style_test.dart]: The TextStyle variables s1 to s5 all depend on each other. Separated what was possible.

File: packages/flutter/lib/src/painting/text_style.dart [hunk 9]
```diff
@@ -122,6 +122,79 @@ import 'basic_types.dart';
...
+/// ### Custom Fonts
+///
+/// Custom fonts can be declared in the `pubspec.yaml` file as shown below:
+///
+///```yaml
+/// flutter:
+///   fonts:
+///     - family: Raleway
+///       fonts:
+///         - asset: assets/fonts/Raleway-Regular.ttf
```
Comment (mit-mit) [inline packages/flutter/lib/src/painting/text_style.dart]: Maybe remove the `assets` part of the path to be consistent with the samples below?

File: packages/flutter/lib/src/painting/text_style.dart [hunk 10]
```diff
@@ -122,6 +122,79 @@ import 'basic_types.dart';
...
+///         - asset: assets/fonts/Raleway-SemiBold.ttf
+///           weight: 600
+///      - family: Schyler
+///        fonts:
+///          - asset: fonts/Schyler-Regular.ttf
+///          - asset: fonts/Schyler-Italic.ttf
+///            style: italic
+///```
+///
+/// The `family` determines the name of the font, which you can use in the
```
Comment (mit-mit) [inline packages/flutter/lib/src/painting/text_style.dart]: For both `family` and `assert` either add `property` after it, or remove the proceeding `the`

File: packages/flutter/lib/src/painting/text_style.dart [hunk 11]
```diff
@@ -122,6 +122,79 @@ import 'basic_types.dart';
...
+/// [fontFamily] argument. The `asset` is a path to the font file, relative to
+/// the [pubspec.yaml] file. The `weight` property specifies the weight of the
+/// glyph outlines in the file as an integer multiple of 100 between 100 and 900.
+/// This corresponds to the [FontWeight] class and can be used in the
+/// [fontWeight] argument. The `style` property specfies whether the outlines in
+/// the file are `italic` or `normal`. These values correspond to the
+/// [FontStyle] class and can be used in the [fontStyle] argument.
+///
+/// To select a custom font create [TextStyle] using the [fontFamily]
```
Comment (mit-mit) [inline packages/flutter/lib/src/painting/text_style.dart]: comma after `font`

**Review by mit-mit — COMMENTED** [review]
Few smaller comments.

Also, I think we may be missing a small update to `pubspec.yaml` in our create template?

**Comment by tvolkert** [comment]
`asset_bundle_package_fonts_test.dart` isn't hermetic and thus will fail when run inside Google's infrastructure.  It's already using the `fs` abstraction, so can you change it to use a MemoryFileSystem instead and ensure that it's fully hermetic?

**Comment by tvolkert** [comment]
Nit: there's also trailing whitespace in `asset_bundle_package_test.dart`

**Comment by szakarias** [comment]
Thanks for the comment. I am using a MemoryFileSystem in the `overrides` of each test .  What am I missing?

**Comment by tvolkert** [comment]
Oops, I missed the function definitions within `main()` and thought the `fs` calls were run in the main method outside the context of a test.  Sorry!

**Comment by szakarias** [comment]
No worries, thanks for looking at it. :-)

**Pull request merged**